Raw version of Waku v2 spec (#144)
vacp2p/rfc: specs/waku/waku-v2.md

---
title: Waku
version: 2.0.0-alpha1
status: Raw
authors: Oskar Thorén <oskar@status.im>
---

## Table of Contents

TODO

## Abstract

> Waku is a privacy-preserving peer-to-peer messaging protocol for resource restricted devices. It implements PubSub in libp2p and adds capabilities on top of it. These capabilities are: (i) retrieving historical messages for mostly-offline devices (ii) adaptive nodes, allowing for heterogenerous nodes to contribute, and (iii) bandwidth preservation for light nodes. This makes it ideal for running a p2p protocol on mobile.

Reviewer: ```suggestion
Waku is a privacy-preserving peer-to-peer messaging protocol for resource restricted devices. It implements PubSub over libp2p and adds capabilities on top of it. These capabilities are: (i) retrieving historical messages for mostly-offline devices (ii) adaptive nodes, allowing for heterogeneous nodes to contribute, and (iii) bandwidth preservation for light nodes. This makes it ideal for running a p2p protocol on mobile.
```

> Historically, it has its roots in Waku v1 [xx5], which stems from Whisper [xx6], originally part of the Ethereum stack. However, Waku v2 acts more as a thin wrapper for PubSub and has a different API. It is implemented in an iterative manner where initial focus is on porting essential functionality to libp2p. See here for a rough road map [xx7].

Reviewer: Maybe link both waku v1 and whisper?

Author: They are linked at the bottom of the document. Will do formatting later, this is just to capture raw state. Lets focus on content and shape of spec first please.

Author: I fixed this anyway now

## Definitions

TODO

## Underlying Transports and Prerequisites

TODO Right now this is more like a set of components

### Peer Discovery

WakuSub and PubSub don't provide an peer discovery mechanism. This has to be provided for by the environment.

### PubSub interface

Waku v2 is implementing the PubSub interface in Libp2p. See PubSub spec [xx2] for more details.

### Protocol Identifier

The current protocol identifier is: `/wakusub/0.0.1` [xx1].

**TODO Update to 2.0.0-alpha0**

### FloodSub

WakuSub is currently a subprotocol of FloodSub. Future versions of WakuSub will support GossipSub 1 [xx3] and GossipSub 1.1 [xx4].

### Bridge mode

To maintain compatibility with Waku v1, a bridge mode can be achieved. See separate spec.

TODO Detail this in a separate spec

## Wire Specification

We are using protobuf RPC messages between peers. Here's what a message looks like, as defined in the PubSub interface.

*NOTE: Should contain protobuf definitions that cover essentials of Waku v1. In cases where it doesn't cover, we can defer to siblings/child specs, e.g. such as the data field for encryption, etc.*

### Messages

```
message Message {

> 	optional string from = 1;

Reviewer 2: Not sure I understand what this is. Why the from?

Author: PubSub spec.

Author: Will explain how we use it (or not) in later commits tho

Reviewer: @oskarth why not just link to the pubsub spec vs copying this in?

	optional bytes data = 2;
	optional bytes seqno = 3;
	repeated string topicIDs = 4;
	optional bytes signature = 5;

> 	optional bytes key = 6;

Reviewer 2: Same as above, why this?

Author: Role TBD, part of PubSub spec

> message Message {
> 	optional string from = 1;
> 	optional bytes data = 2;
> 	optional bytes seqno = 3;
> 	repeated string topicIDs = 4;
> 	optional bytes signature = 5;
> 	optional bytes key = 6;
> }

Reviewer: ```suggestion
message Message {
	optional string from = 1;
	optional bytes data = 2;
	optional bytes sequence_number = 3;
	repeated string topic_ids = 4;
	optional bytes signature = 5;
	optional bytes key = 6;
}
```

Because the keys in protobuf aren't encoded, firstly let's use meaningful names. Additionally, use `snake case` because it encodes it properly when it comes to the language.

Author: This is from the PubSub spec. Let's not edit this IMO.

Author: I mean I suppose we could, it just seems like it could get confusing when people compare specs.

What are people's thoughts?

- [Copyright](#copyright)
- [References](#references)

## Abstract

Waku is a privacy-preserving peer-to-peer messaging protocol for resource
restricted devices. It implements PubSub over libp2p and adds capabilities on
top of it. These capabilities are: (i) retrieving historical messages for
mostly-offline devices (ii) adaptive nodes, allowing for heterogeneous nodes to
contribute, and (iii) bandwidth preservation for light nodes. This makes it
ideal for running a p2p protocol on mobile.

Historically, it has its roots in [Waku v1](specs.vac.dev/waku/waku.html), which
stems from [Whisper](https://eips.ethereum.org/EIPS/eip-627), originally part of
the Ethereum stack. However, Waku v2 acts more as a thin wrapper for PubSub and
has a different API. It is implemented in an iterative manner where initial
focus is on porting essential functionality to libp2p. See [rough road
map](https://vac.dev/waku-v2-plan).

## Motivation and goals

1. **Generalized messaging.** Many applications requires some form of messaging
   protocol to communicate between different subsystems or different nodes. This
   messaging can be human-to-human or machine-to-machine or a mix.

> 2. **Peer-to-peer.**These applications sometimes have requirements that make

Reviewer: ```suggestion
2. **Peer-to-peer.** These applications sometimes have requirements that make
```

   them suitable for peer-to-peer solutions.

3. **Resource restricted**.These applications often run in constrained
   environments, where resources or the environment is restricted in some
   fashion. E.g.:

- limited bandwidth, CPU, memory, disk, battery, etc
- not being publicly connectable
- only being intermittently connected; mostly-offline

4. **Privacy.** These applications have a desire for some privacy guarantees,
   such as pseudonymity, metadata protection in transit, etc.

Waku provides a solution that satisfies these goals in a reasonable matter.

## Definitions

TODO

## Underlying Transports and Prerequisites

TODO Right now this is more like a set of components

### Peer Discovery

WakuSub and PubSub don't provide an peer discovery mechanism. This has to be
provided for by the environment.

### PubSub interface

Waku v2 is implementing the PubSub interface in Libp2p. See [PubSub interface

> for libp2p (r2,
> 2019-02-01)](https://github.com/libp2p/specs/blob/master/pubsub/README.md) for

Reviewer: ```suggestion
for libp2p (r2, 2019-02-01)](https://github.com/libp2p/specs/blob/master/pubsub/README.md) 
for
```

more details.

### Protocol Identifier

The current [protocol identifier](https://docs.libp2p.io/concepts/protocols/)
is: `/wakusub/2.0.0-alpha1`.

### FloodSub

WakuSub is currently a subprotocol of FloodSub. Future versions of WakuSub will

> support [GossipSub
> v1.0](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md)
> and [GossipSub
> 1.1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md).

Reviewer: ```suggestion
support [GossipSub v1.0](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md)
and [GossipSub 1.1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md).
```
